pkg/inlinekeepalive/keepalive.go

```diff
@@ -0,0 +1,84 @@
+package inlinekeepalive
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/hashicorp/go-hclog"
+	"google.golang.org/protobuf/reflect/protoreflect"
+
+	pb "github.com/hashicorp/waypoint/pkg/server/gen"
+)
+
+const (
+	KeepaliveProtoSignature   = "inline_keepalive"
+	HeaderSendKeepalivesKey   = "wp-inline-keepalives"
+	HeaderSendKeepalivesValue = "true"
+	FeatureName               = "inline-keepalives"
+)
+
+// IsInlineKeepalive determines if a given proto message is
+// an inline keepalive.
+func IsInlineKeepalive(log hclog.Logger, m protoreflect.ProtoMessage) bool {
+	unknownFields := m.ProtoReflect().GetUnknown()
+	if unknownFields == nil {
+		// No unknown fields here, not our keepalive, continue as normal
+		return false
+	}
+
+	keepAlive := pb.InlineKeepalive{}
+	err := proto.Unmarshal(unknownFields, &keepAlive)
+	if err != nil {
+		// couldn't marshal to the keepalive message, continue as normal
+		return false
+	}
+
+	if keepAlive.Signature != KeepaliveProtoSignature {
+		// We had some other protobuf message with an unknown field whose number matched the high-order
+		// keepalive message field number. This seems very unlikely.
+		log.Warn("Unexpectedly received proto message with an unknown field matching the inlinekeepalive reserved field number")
```

Comment:
Can we include the field as a key=>val in the logger? Would be nice to see what it is for extra debug info if we ever happen to reach this

Reply:
I did a quick pass, and I didn't see a way to get at the field number from here. I can hard-code it, which doesn't feel too bad, as it's hard-coded into server.proto. Do you think that would help?

Reply:
Oh I see. I was mostly thinking if the signature we're receiving is variable it would be nice to know but if it's always the same then no biggie!